controller/userController.js

```diff
@@ -86,7 +92,180 @@ const token = (req, res) => {
+        } else {
+            return res.status(HttpStatus.NOT_FOUND).json({
+                status: 'Error',
+                message: 'Users not found.'
+            });
+        }
+    }).catch(err => {
+        const errorMessage = ValidationUtil.buildErrorMessage(err, 'retrieveAll', 'user');
+        return res.status(HttpStatus.INTERNAL_SERVER_ERROR).json({
+            status: 'Error',
+            message: errorMessage
+        });
+    });
+};
+
+const retrieveById = (req, res) => {
+    const userId = req.params.id;
+    if (userId.length != AppConfig.OBJECT_ID_LEN) {
+        return res.status(HttpStatus.BAD_REQUEST).json({
+            status: 'Error',
+            message: 'Invalid ID'
+        });
+    }
+
+    User.findById(userId, { passwordHash: 0, salt: 0 }).then(user => {
+        if (user) {
+            return res.status(HttpStatus.OK).json(user);
+        } else {
+            return res.status(HttpStatus.NOT_FOUND).json({
+                status: 'Error',
+                message: 'User not found.'
+            });
+        }
+    }).catch(err => {
+        const errorMessage = ValidationUtil.buildErrorMessage(err, 'retrieveById', 'user');
+        return res.status(HttpStatus.INTERNAL_SERVER_ERROR).json({
+            status: 'Error',
+            message: errorMessage
+        });
+    });
+};
+
+const update = (req, res) => {
+    const userId = req.params.id;
+    if (userId.length != AppConfig.OBJECT_ID_LEN) {
+        return res.status(HttpStatus.BAD_REQUEST).json({
+            status: 'Error',
+            message: 'Invalid ID'
+        });
+    }
+    user = {};
+    // Update fields
+    if (req.body.name)
+        user.name = req.body.name;
+    if (req.body.surname)
+        user.surname = req.body.surname;
+    if (req.body.email)
+        user.email = req.body.email;
+    if (req.body.dateOfBirth)
+        user.dateOfBirth = req.body.dateOfBirth;
+    if (req.body.active)
+        user.active = req.body.active;
+    if (req.body.passwordHash) {
+        let retVal = getPasswordHash(req.body.passwordHash);
+        user.salt = retVal.salt;
+        user.passwordHash = retVal.passwordHash;
+    }
+
+    User.findOneAndUpdate({ _id: userId }, user, { useFindAndModify: false, new: true, runValidators: true }).then(newUser => {
+        if (newUser) {
+            newUser.passwordHash = null;
+            newUser.salt = null;
+            return res.status(HttpStatus.OK).json(newUser);
+        } else {
+            return res.status(HttpStatus.NOT_FOUND).json({
+                status: 'Error',
+                message: 'User not found.'
+            });
+        }
+    }).catch(err => {
+        const errorMessage = ValidationUtil.buildErrorMessage(err, 'update', 'user');
+        return res.status(HttpStatus.INTERNAL_SERVER_ERROR).json({
+            status: 'Error',
+            message: errorMessage
+        });
+    });
+};
+
+const remove = (req, res) => {
```

Comment:
Again. CompanyID check, Super-Admin/Admin check is necessary. This makes it seem that anyone can remove anyone.

Reply:
Admin check will be implemented in Auth middleware.